### PR TITLE
Meta transactions

### PR DESCRIPTION
**Pull request by mrsmkl**
## Description

Adding support for meta transactions. To avoid adding a big dependency with web3.js and truffle stuff back, there will juts be a signer that can be passed in tx parameters.

## Is this PR related with an open issue?

Related to Issue #

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [ ] Follows the code style of this project.
- [ ] Tests Cover Changes
- [ ] Documentation

#### Funny gif

Put a link of a funny gif inside the parenthesis-->

### CODE REVIEW

**Review by aaitor — COMMENTED** [review]
looking good

File: integration/nevermined/MetaTx.test.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                BigNumber.from(await nevermined.keeper.nftUpgradeable.balance(await wallet.getAddress(), ddo.id)),
+                BigNumber.from(2)
+            )
+
+            const meta: MetaTxParameters = {
+                wallet,
+                paymasterAddress
+            }
+            await nevermined.keeper.didRegistry.burn(ddo.id, BigNumber.from(2), await wallet.getAddress(), {meta})
+            // await nevermined.nfts.burn(ddo.id, BigNumber.from(6), artist, {meta})
```
Comment (r-marques) [inline integration/nevermined/MetaTx.test.ts:94]: cleanup

File: src/keeper/contracts/ContractBase.ts [hunk 2]
```diff
@@ -1,24 +1,56 @@
...
+    provider: ethers.providers.JsonRpcProvider
+    constructor(provider: ethers.providers.JsonRpcProvider) {
+        this.provider = provider
+    }
+    send(payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void): void {
+        const id = typeof payload.id === 'string' ? parseInt(payload.id) : payload.id
+        this.provider.send(payload.method, payload.params).then(result => callback(null, {jsonrpc: payload.jsonrpc, id, result}))
+    }
+}
+
```
Comment (r-marques) [inline src/keeper/contracts/ContractBase.ts:53]: I'm assuming that all this relates to opengsn. If so this should have better names and maybe be on a separate source file

File: tsconfig.json [hunk 3]
```diff
@@ -3,6 +3,7 @@
     "resolveJsonModule": true,
     "moduleResolution": "Node",
     "esModuleInterop": true,
+    "skipLibCheck": true,
```
Comment (r-marques) [inline tsconfig.json:6]: what is the reason for this change?

File: integration/nevermined/MetaTx.test.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                BigNumber.from(8)
+            )
+            assert.deepEqual(
+                await nevermined.nfts.balance(ddo.id, collector),
+                BigNumber.from(2)
+            )
+        })
+
+        it('metatransactions should work', async () => {
+            const wallet = new ethers.Wallet(Buffer.from('1'.repeat(64), 'hex'))
```
Comment (r-marques) [inline integration/nevermined/MetaTx.test.ts:81]: maybe just `wallet.createRandom()`

**Comment by r-marques** [comment]
also `yarn.lock` is part of `.gitignore`. It shouldn't be committed unless we have a reason for that

File: tsconfig.json (hunk 3 above)
Comment (mrsmkl) [reply tsconfig.json:6]: Looks like the current "stable" version of opengsn is kind of broken, here's what happens without it https://forum.opengsn.org/t/typescript-compiler-fails-with-opengsn-v2-2-1/169

File: src/keeper/contracts/conditions/NFTs/NFTUpgradable.ts [hunk 5]
```diff
@@ -103,7 +103,7 @@ export class NFTUpgradeable extends ContractBase {
      * @returns
      */
     public async balance(address: string, did: string): Promise<BigNumber> {
-        return this.call('balanceOf', [zeroX(address), didZeroX(did)])
+        return BigNumber.from(await this.call('balanceOf', [zeroX(address), didZeroX(did)]) || 0)
```
Comment (r-marques) [inline src/keeper/contracts/conditions/NFTs/NFTUpgradable.ts:106]: I would remove `|| 0`. the call needs to return a number

Comment (mrsmkl) [reply src/keeper/contracts/conditions/NFTs/NFTUpgradable.ts:106]: yeah, I just added it because I forgot to add `await` first

File: .github/workflows/testing-new-gateway.yml [hunk 6]
```diff
@@ -73,8 +73,11 @@ jobs:
 
           # Disabling the compute api for now
           # Check https://github.com/nevermined-io/sdk-js/issues/33
-
-          ./start_nevermined.sh --latest ${{ matrix.network }} --no-graph --no-gateway --opengsn 2>&1 | tee nevermined_tools.txt &
+          
+          # TODO: Use the default Node when merged into master
+          export KEEPER_VERSION=latest
+          export NODE_VERSION=latest
```
Comment (r-marques) [inline .github/workflows/testing-new-gateway.yml:79]: aren't we already using latest by default?
also I don't think we need to `NODE_VERSION` since the nevermined tools is started with `--no-node`

File: .github/workflows/testing-nightly.yml [hunk 7]
```diff
@@ -48,7 +48,7 @@ jobs:
           # Disabling the compute api for now
           # Check https://github.com/nevermined-io/sdk-js/issues/33
 
-          ./start_nevermined.sh --latest ${{ matrix.network }} --opengsn --new-gateway 2>&1 | tee nevermined_tools.txt &
+          ./start_nevermined.sh --latest ${{ matrix.network }} --opengsn --new-node 2>&1 | tee nevermined_tools.txt &
```
Comment (r-marques) [inline .github/workflows/testing-nightly.yml:51]: `--new-node` is the default so it's no needed here

File: .github/workflows/testing.yml [hunk 8]
```diff
@@ -47,7 +47,10 @@ jobs:
           # Disabling the compute api for now
           # Check https://github.com/nevermined-io/sdk-js/issues/33
 
-          ./start_nevermined.sh --latest ${{ matrix.network }} --opengsn --new-gateway 2>&1 | tee nevermined_tools.txt &
+          # TODO: Use the default Node when merged into master
+          export KEEPER_VERSION=latest
+          export NODE_VERSION=latest
```
Comment (r-marques) [inline .github/workflows/testing.yml:52]: same as before

File: src/keeper/contracts/ContractBase.ts [hunk 9]
```diff
@@ -10,8 +10,11 @@ import { ContractReceipt, ethers } from 'ethers'
 import BigNumber from '../../utils/BigNumber'
 
 export interface MetaTxParameters {
+    /*
     paymasterAddress: string
     wallet: ethers.Wallet
+    */
```
Comment (r-marques) [inline src/keeper/contracts/ContractBase.ts:16]: cleanup

File: src/keeper/contracts/ContractBase.ts [hunk 10]
```diff
@@ -210,19 +187,9 @@ export abstract class ContractBase extends Instantiable {
...
-            params.meta = undefined
-            const gsnProvider = RelayProvider.newProvider({ provider: new Web3ProviderWrapper(this.web3), config })
-            await gsnProvider.init()
-            gsnProvider.addAccount(wallet.privateKey)
-            const etherProvider = new ethers.providers.Web3Provider(gsnProvider)
-            const contract = this.contract.connect(etherProvider.getSigner(wallet.address))
-            return await this.internalSend(name, from, args, params, params, contract)
+            const paramsFixed = {...params, meta: undefined}
+            const contract = this.contract.connect(params.meta.signer)
+            return await this.internalSend(name, from, args, paramsFixed, paramsFixed, contract)
```
Comment (r-marques) [inline src/keeper/contracts/ContractBase.ts:192]: I think internalSend has a duplicate parameter

File: src/keeper/utils.ts [hunk 11]
```diff
@@ -54,3 +55,29 @@ export async function getNetworkName(networkId: number): Promise<string> {
...
+export class Web3ProviderWrapper {
+    provider: ethers.providers.JsonRpcProvider
+    constructor(provider: ethers.providers.JsonRpcProvider) {
+        this.provider = provider
+    }
+    send(payload: JsonRpcPayload, callback: (error: Error | null, result?: JsonRpcResponse) => void): void {
+        const id = typeof payload.id === 'string' ? parseInt(payload.id) : payload.id
+        this.provider.send(payload.method, payload.params).then(result => callback(null, {jsonrpc: payload.jsonrpc, id, result}))
+    }
+}
```
Comment (r-marques) [inline src/keeper/utils.ts:83]: is this still needed? it doesn't seem to be used anywhere

Comment (mrsmkl) [reply src/keeper/utils.ts:83]: it will be useful when somebody actually wants to use the meta transactions, but currently it's only used in the test

**Review by r-marques — APPROVED** [review]
